lib/backend/k8s/resources/ippools.go

```diff
@@ -0,0 +1,177 @@
+	return kvp, nil
+}
+
+func (c *client) Update(kvp *model.KVPair) (*model.KVPair, error) {
+	tpr := IPPoolToThirdParty(kvp)
+	res := thirdparty.IpPool{}
+	req := c.tprClient.Put().
+		Resource("ippools").
+		Namespace("kube-system").
+		Body(tpr).
+		Name(tpr.Metadata.Name)
+	err := req.Do().Into(&res)
+	if err != nil {
+		return nil, utils.K8sErrorToCalico(err, kvp.Key)
+	}
+	kvp.Revision = tpr.Metadata.ResourceVersion
+	return kvp, nil
+}
+
+func (c *client) Apply(kvp *model.KVPair) (*model.KVPair, error) {
+	updated, err := c.Update(kvp)
+	if err != nil {
+		if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+			return nil, err
+		}
+
+		// It doesn't exist - create it.
+		updated, err = c.Create(kvp)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return updated, nil
+}
+
+func (c *client) Delete(kvp *model.KVPair) error {
+	result := c.tprClient.Delete().
+		Resource("ippools").
+		Namespace("kube-system").
+		Name(tprName(kvp.Key.(model.IPPoolKey))).
+		Do()
+
+		// TODO: Error conversion
+	return utils.K8sErrorToCalico(result.Error(), kvp.Key)
+}
+
+func (c *client) Get(key model.Key) (*model.KVPair, error) {
+	tpr := thirdparty.IpPool{}
+	err := c.tprClient.Get().
+		Resource("ippools").
+		Namespace("kube-system").
+		Name(tprName(key.(model.IPPoolKey))).
+		Do().Into(&tpr)
+	if err != nil {
+		return nil, utils.K8sErrorToCalico(err, key)
+	}
+
+	return ThirdPartyToIPPool(&tpr), nil
+}
+
+func (c *client) List(list model.ListInterface) ([]*model.KVPair, error) {
+	kvps := []*model.KVPair{}
+	tprs := thirdparty.IpPoolList{}
+	l := list.(model.IPPoolListOptions)
+
+	// Build the request.
+	req := c.tprClient.Get().Resource("ippools").Namespace("kube-system")
+	if l.CIDR.IP != nil {
+		k := model.IPPoolKey{CIDR: l.CIDR}
+		req.Name(tprName(k))
+	}
+
+	// Perform the request.
+	err := req.Do().Into(&tprs)
+	if err != nil {
+		// Don't return errors for "not found".  This just
+		// means there are no IPPools, and we should return
+		// an empty list.
+		if !kerrors.IsNotFound(err) {
+			return nil, utils.K8sErrorToCalico(err, l)
+		}
+	}
+
+	// Convert them to KVPairs.
+	for _, tpr := range tprs.Items {
+		kvps = append(kvps, ThirdPartyToIPPool(&tpr))
+	}
+	return kvps, nil
+}
+
+func (c *client) EnsureInitialized() error {
+	log.Info("Ensuring IP Pool ThirdPartyResource exists")
+	tpr := extensions.ThirdPartyResource{
+		ObjectMeta: kapiv1.ObjectMeta{
+			Name:      "ip-pool.projectcalico.org",
+			Namespace: "kube-system",
+		},
+		Description: "Calico IP Pools",
+		Versions:    []extensions.APIVersion{{Name: "v1"}},
+	}
+	_, err := c.clientSet.Extensions().ThirdPartyResources().Create(&tpr)
+	if err != nil {
+		// Don't care if it already exists.
+		if !kerrors.IsAlreadyExists(err) {
+			return goerrors.New(fmt.Sprintf("failed to create ThirdPartyResource %s: %s", tpr.ObjectMeta.Name, err))
+		}
+	}
+	return nil
+}
+
+// Syncer is not implemented for ip pools.
+func (c *client) Syncer(callbacks api.SyncerCallbacks) api.Syncer {
```

Comment:
I considered that briefly.

Unfortunately, I think the way the Syncer API works means it would be a lot of work to do this and would require the other resources to also implement.

Ultimately, I'd like to add a `Watch()` method to the Client interface which can be implemented by the resources.